disputer/disputer.js

```diff
@@ -0,0 +1,84 @@
+const { Logger } = require("../financial-templates-lib/Logger");
+
+class Disputer {
+  constructor(expiringMultiPartyClient, account) {
+    this.account = account;
+
+    // Expiring multiparty contract to read contract state
+    this.empClient = expiringMultiPartyClient;
+
+    // Instance of the expiring multiparty to perform on-chain liquidations
+    this.empContract = this.empClient.emp;
+  }
+
+  // Queries disputable liquidations and disputes any that were incorrectly liquidated.
+  queryAndDispute = async priceFunction => {
+    Logger.debug({
+      at: "Disputer",
+      message: "Checking for any disputable liquidations"
+    });
+
+    // Update the client to get the latest liquidation information.
+    await this.empClient._update();
+
+    // Get the latest disputable liquidations from the client.
+    const undisputedLiquidations = this.empClient.getUndisputedLiquidations();
+    const disputeableLiquidations = undisputedLiquidations.filter(liquidation =>
+      this.empClient.isDisputable(liquidation, priceFunction(liquidation.liquidationTime))
+    );
+
+    if (disputeableLiquidations.length === 0) {
+      Logger.debug({
+        at: "liquidator",
+        message: "No disputable liquidations"
+      });
+
+      // Nothing left to do, so return.
+      return;
+    }
+
+    Logger.info({
+      at: "Disputer",
+      message: "Disputable liquidation(s) detected!",
+      number: disputeableLiquidations.length,
+      underCollateralizedPositions: disputeableLiquidations
+    });
+
+    for (const disputeableLiquidation of disputeableLiquidations) {
+      Logger.info({
+        at: "Disputer",
+        message: "Disputing liquidation",
+        address: disputeableLiquidation.sponsor,
+        inputPrice: priceFunction(disputeableLiquidation.liquidationTime)
+      });
+
+      // Create the liquidation transaction
+
+      // TODO: compute the amount of collateral required to dispute and ensure the bot has enough to perform it.
+
+      // TODO: handle transaction failures.
+      const receipt = await this.empContract.methods
```

Comment:
Doing it like this is a blocking transaction. this is fine but it means that a bunch of disputes will wait until they mine before sending out the following transaction. Do you want to make this the same as how it is in the [liquidator](https://github.com/UMAprotocol/protocol/blob/63a6daec192ddfe585d225c3164a4a09a43c5adf/liquidator/liquidator.js#L69) with `promise.all` or you think it's best like this for now?

Reply:
Yeah, I thought you tested and determined that this doesn't actually parallelize the transactions because there's a blocking step before the send? Furthermore, I was thinking more about it, and if we try to do these in parallel, it may make collateral/inventory management more complex. For those reasons, I I just decided to go with the simpler approach until we determine that we can and need to do the parallelization.

Reply:
Nope. I spent a bunch of time last week testing it out and it worked perfectly well with the `promise.all`. I used ganache with a normal block time and was able to submit ~20 transactions in one non-blocking go because they all enter the mem-pool in parallel and then resolve as they finish.

However, your point on collateral/inventory management does become more complex. I'm happy to leave it like this for now but this does not scale to dealing with many sponsors being liquidated/disputed at a high frequency or in times of network stress.

Reply:
SGTM. I say we just keep it this way for simplicity. We can move to batching them later. There are a few other experiments I want to run to ensure the batching actually does what we think it does, especially on public networks. We can save that for later, though.

Reply:
But your point is well taken that this may be inefficient.

Reply:
no problem. let's keep it simple for now 👍 .